File: cogs/voice.py
import discord
import datetime

from discord.ext import tasks, commands
from discord.ext.commands.errors import MemberNotFound
from discordtools import WaitFor
from tools.database import db
from views.confirmation import ConfirmationView
from sqlite3 import IntegrityError

class VoiceChannel:
    def __init__(self, bot):
        self.bot = bot
        self.voice_channel = None
        self.text_channel = None
        self.guild = None
        self.creator = None
        self.delete_time = None
        self.whitelist = False
        self.people = []

    @staticmethod
    def get_all(bot):
        voices = []
        datas = db.get_voice()
        for data in datas:
            voice = VoiceChannel(bot)
            voice.load(voiceID=data['voiceID'])
            voices.append(voice)
        return voices

    def load(self, **options):
        data = db.get_voice(**options)[0]
        if not data:
            raise ValueError('No Voice')
        self.voice_channel = self.bot.get_channel(int(data['voiceID']))
        self.guild = self.voice_channel.guild
        self.text_channel = self.guild.get_channel(int(data['textID']))
        self.creator = self.guild.get_member(int(data['creator']))
        self.delete_time = data['deleteTime']
        self.whitelist = bool(data['whitelist'])
        self.people = (self.guild.get_member(int(x)) for x in data['people'].split(','))

    async def new(self, guild, name, creator: discord.Member, whitelist: bool, people: list):
        MINUTES_TO_DELETE = 10

        self.guild = guild
        self.creator = creator
        self.whitelist = whitelist
        self.delete_time = datetime.datetime.now() + datetime.timedelta(minutes=MINUTES_TO_DELETE)
        self.people = people
        private_category = None # Get Category
        for category in guild.categories:
            if category.name == 'Private Channels':
                private_category = category

        allow_permissions = discord.PermissionOverwrite( # Set Permissions
            view_channel = True,
            connect = True
        )
        deny_permissions = discord.PermissionOverwrite(
            view_channel = False,
            connect = False
        )
        if whitelist: # Build overwrites dict for discord
            overwrites = {
                guild.default_role: deny_permissions
            }
            for p in people:
                overwrites[p] = allow_permissions
        else:
            overwrites = {
                guild.default_role: allow_permissions
            }
            for p in people:
                overwrites[p] = deny_permissions
        if private_category: # Choose whether to make it in category or guild
            c = private_category
        else:
            c = guild
        topic = f'Private text channel coupled with the voice channel below.'
        self.text_channel = await c.create_text_channel(
            name,
            overwrites=overwrites,
            topic=topic,
            reason=f'Created by {creator}'
        )
        self.voice_channel = await c.create_voice_channel(
            name,
            overwrites=overwrites,
            reason=f'Created by {creator}'
        )

        embed = discord.Embed(
            title=f'{name} Created Successfully',
            description=(
                f'This channel will persist for {MINUTES_TO_DELETE} minutes or until everyone is gone, whichever comes last.\n'
                'Keep in mind, Administrators can still see all of these channels.'
            ),
            colour=discord.Colour.green()
        )
        embed.add_field(
            name='Creator',
            value=f'{self.creator.mention}'
        )
        people_string = '\n'.join(e.mention for e in self.people)
        embed.add_field(
            name=f'{"Allowed" if self.whitelist else "Blocked"} People',
            value=people_string if people_string else 'None'
        )
        await self.text_channel.send(embed=embed)
        db.add_voice(
            self.creator.id,
            self.voice_channel.id,
            self.text_channel.id,
            self.whitelist,
            filter(lambda x: bool(x), (x.id if x else x for x in self.people)),
            self.delete_time
        )
        try:
            db.add_last(self.creator.id)
        except IntegrityError:
            pass
        db.set_last(self.creator.id, name, self.whitelist, filter(lambda x: bool(x), (x.id if x else x for x in self.people)))

    async def check_delete(self):
        if not self.voice_channel.members and datetime.datetime.now() > self.delete_time:
            await self.delete()

    async def delete(self):
        db.delete_voice(self.voice_channel.id)
        try:
            await self.text_channel.delete()
        except (discord.NotFound, discord.Forbidden, AttributeError):
            pass
        try:
            await self.voice_channel.delete()
        except (discord.NotFound, discord.Forbidden, AttributeError):
            pass


class Voice(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.voice_checker.start()

    @tasks.loop(seconds=10)
    async def voice_checker(self):
        voices = VoiceChannel.get_all(self.bot)
        for voice in voices:
            await voice.check_delete()

    @voice_checker.before_loop
    async def before_voice_checker(self):
        await self.bot.wait_until_ready()

    @commands.group()
    @commands.guild_only()
    async def voice(self, ctx):
        """Use -help voice for more info."""
        pass

    @commands.Cog.listener()
    async def on_message(self, message: discord.Message):
        if message.guild.me in message.mentions:
            member = message.author
            if member.voice:
                vc = member.voice.channel
                if vc:
                    embed = discord.Embed(
                        title=f'Mention Everyone in {vc.name}?',
                        colour=discord.Colour.orange()
                    )
                    view = ConfirmationView()
                    members_to_mention = vc.members
                    q_message = await message.channel.send(embed=embed, view=view)
                    confirmed = await view.wait_for_answer()
                    if confirmed:
                        await q_message.delete()
                        await message.channel.send(
                            f'{member.mention} pinged:\n' + ' '.join(e.mention for e in members_to_mention)
                        )
                    else:
                        await q_message.delete()

    @voice.command()
    @commands.max_concurrency(1, per=commands.BucketType.channel)
    async def new(self, ctx):
        """Create a new private channel with the desired settings."""
        if db.get_voice(creator=ctx.author.id):
            embed = discord.Embed(
                title='Do you really need another one?',
                colour=discord.Colour.red()
            )
            await ctx.send(embed=embed)
            return
        wait = WaitFor(self.bot, 'message', lambda message: message.author == ctx.author and message.channel == ctx.channel, 30)
        wait.set_messageable(ctx)
        embed = discord.Embed(
            title="Channel Setup",
            colour=discord.Colour.green()
        )
        wait.set_embed(embed)
        last = db.get_last(ctx.author.id)
        redo = True
        make = False
        stage = 0
        
        while redo:
            if last and stage == 0:
                embed.description = (
                    'Would you like to create a `new` configuration or use your `last` one?'
                )
                embed.set_footer(text='Reply with either: new/last')
                value = await wait.run()
                if value:
                    content = value.content
                else:
                    stage = -1
                    continue
                if content == 'new':
                    stage = 1
                elif content == 'last':
                    stage = -2
                else:
                    stage = -1
            elif stage == 0:
                stage = 1

            if stage == 1:
                embed.description = 'What would you like to name the new channel?'
                embed.set_footer(text='Reply with a name between 1 and 100 characters')
                value = await wait.run()
                if value:
                    name = value.content
                    length = len(name)
                    if length > 1 and length < 100:
                        stage = 2
                    else:
                        warning = await wait.messageable.send('Name MUST be within 1 and 100 Characters.')
                        wait.bot_messages.append(warning)
                else:
                    stage = -1
                    continue
            elif stage == 2:
                embed.description = 'Would you like the list of players that you will give in the next step to be a blacklist or a whitelist?'
                embed.set_footer(text='Reply with either: blacklist/whitelist')
                value = await wait.run()
                if value:
                    content = value.content.lower()
                    if content == 'blacklist':
                        whitelist = False
                        stage = 3
                    elif content == 'whitelist':
                        whitelist = True
                        stage = 3
                    else:
                        warning = await wait.messageable.send('Invalid response, if you wish to cancel just wait for half a minute.')
                        wait.bot_messages.append(warning)
                else:
                    stage = -1
                    continue
            elif stage == 3:
                embed.description = f'Enter a space separated list of users to {"whitelist" if whitelist else "blacklist"}.'
                embed.set_footer(text='Reply with a space separated list of user OR reply with 0 to have an empty list')
                value = await wait.run()
                list_of_members = []
                if value:
                    if value.content != '0':
                        content_list = value.content.split(' ')
                        converter = commands.MemberConverter()
                        for raw in content_list:
                            try:
                                result = await converter.convert(ctx, raw)
                            except MemberNotFound:
                                result = None
                            if result:
                                list_of_members.append(result)
                    make = True
                    stage = -1
                else:
                    stage = -1
                    continue
            elif stage == -2:
                await self.last(ctx)
                stage = -1
            elif stage == -1:
                redo = False
        
        if make:
            voice = VoiceChannel(self.bot)
            people = list(list_of_members)
            if whitelist:
                people.append(ctx.author)
            await ctx.message.delete()
            await voice.new(ctx.guild, name, ctx.author, whitelist, people)
        await wait.cleanup()

    @voice.command()
    async def delete(self, ctx):
        """Use this in the text channel of what you want to delete."""
        voice = VoiceChannel(self.bot)
        voice.load(textID=ctx.channel.id)
        if ctx.author == voice.creator or ctx.author.guild_permissions.manage_channels:
            await voice.delete()
        else:
            await ctx.send('You do not have the appropriate permissions to delete this.')

    @voice.command()
    async def last(self, ctx):
        """Create a voice channel based on the last settings used."""
        if db.get_voice(creator=ctx.author.id):
            embed = discord.Embed(
                title='Do you really need another one?',
                colour=discord.Colour.red()
            )
            await ctx.send(embed=embed)
            return
        people = []
        last = db.get_last(ctx.author.id)
        if not last:
            return
        if last['people']:
            for p in last['people'].split(','):
                i = int(p)
                m = ctx.guild.get_member(i)
                people.append(m)
        voice = VoiceChannel(self.bot)
        await ctx.message.delete()
        await voice.new(ctx.guild, last['name'], ctx.author, last['whitelist'], people)

    @voice.command()
    @commands.cooldown(1, 900, type=commands.BucketType.member)
    async def current(self, ctx, name: str):
        """Creates a whitelisted voice channel based on the members in the current vc."""
        if db.get_voice(creator=ctx.author.id):
            embed = discord.Embed(
                title='Do you really need another one?',
                colour=discord.Colour.red()
            )
            await ctx.send(embed=embed)
            return
        channel = ctx.author.voice.channel
        if not channel:
            return
        voice = VoiceChannel(self.bot)
        await ctx.message.delete()
        await voice.new(ctx.guild, name, ctx.author, True, list(channel.members))
        for m in channel.members:
            try:
                await m.move_to(voice.voice_channel, reason='New Channel Created')
            except discord.HTTPException:
                pass

def setup(bot):
    bot.add_cog(Voice(bot))

File: models/voice.js
const { Voice: VoiceDB, VoiceTemplate } = require('../tools/database');
const { GuildMember, MessageEmbed, Collection } = require('discord.js');
const { userMention } = require('@discordjs/builders');

class Voice {
    /**
     * @param {string} name
     * @param {GuildMember} member
     * @param {string} voiceChannelId
     * @param {string} textChannelId
     */
    constructor(server, voiceChannelId = null, textChannelId = null) {
        this.server = server;
        this.name = '';
        this.creator = null;
        this.guild = server.guild;
        this.voiceChannelId = voiceChannelId;
        this.textChannelId = textChannelId;
        this.whitelist = false;
        this.members = new Collection();
        this.creationTimestamp = Number(Date.now());
        this.welcomeMessage = null;
        this.loaded = false;

        if (voiceChannelId) {
            this.voiceChannel = this.guild.channels.cache.get(voiceChannelId);
        } else {
            this.voiceChannel = null;
        }
        if (textChannelId) {
            this.textChannel = this.guild.channels.cache.get(textChannelId);
        } else {
            this.textChannel = null;
        }
    }

    static MINUTES = 10;

    toJSON() {
        return {
            creatorId: this.creator.id,
            voiceId: this.voiceChannelId,
            textId: this.textChannelId,
            whitelist: this.whitelist,
            welcomeId: this.welcomeMessage ? this.welcomeMessage.id : null,
            members: [...this.members.values()].map(m => m.id),
        };
    }

    async build(name, member, whitelist, nsfw = false) {
        this.name = name;
        this.creator = member;
        this.whitelist = whitelist;
        const data = await VoiceDB.findOne({ where: { creatorId: this.creator.id } });
        if (data) return false;
        const privateCategory = this.server.privateCategory;
        if (!privateCategory) return false;
        const permissions = [];
        if (whitelist) {
            this.members.set(this.guild.client.user.id, this.guild.client.user);
            this.members.set(member.id, member);
        }
        for (const id of this.members.keys()) {
            if (this.whitelist) {
                const allowPermission = Voice.getAllowPermission(id);
                permissions.push(allowPermission);
            } else {
                const denyPermission = Voice.getDenyPermission(id);
                permissions.push(denyPermission);
            }
        }
        // Set everyone permissions
        if (!this.whitelist) {
            const allowPermission = Voice.getAllowPermission(this.guild.roles.everyone.id, 'role');
            permissions.push(allowPermission);
        } else {
            const denyPermission = Voice.getDenyPermission(this.guild.roles.everyone.id, 'role');
            permissions.push(denyPermission);
        }

        // Create channels
        const voiceChannel = await this.guild.channels.create(this.name, {
            type: 'GUILD_VOICE',
            parent: privateCategory,
            permissionOverwrites: permissions,
            reason: `Created by ${this.creator.user.tag}`,
        });
        const textChannel = await this.guild.channels.create(this.name, {
            type: 'GUILD_TEXT',
            parent: privateCategory,
            permissionOverwrites: permissions,
            reason: `Created by ${this.creator.user.tag}`,
            nsfw: nsfw,
        });
        this.textChannel = textChannel;
        this.textChannelId = textChannel.id;
        this.voiceChannel = voiceChannel;
        this.voiceChannelId = voiceChannel.id;
        this.creationTimestamp = voiceChannel.createdTimestamp;
        this.loaded = true;

        // Send welcome message
        this.welcomeMessage = await this.textChannel.send({ embeds: this.getEmbeds() });
        await this.welcomeMessage.pin();
        await VoiceDB.create(this.toJSON());
        this.getVoiceTemplate().save();
    }

    async edit(data = { name: undefined, whitelist: undefined }) {
        if (data.name) this.name = data.name;

        const overwrites = [];
        if (data.whitelist !== undefined && data.whitelist !== this.whitelist) {
            this.whitelist = data.whitelist;
            if (this.whitelist) {
                this.members.set(this.guild.client.user.id, this.guild.client.user);
                this.members.set(this.creator.id, this.creator);
            } else {
                this.members.delete(this.guild.client.user.id);
                this.members.delete(this.creator.id);
            }
            for (const member of this.members.values()) {
                overwrites.push(this.getPermission(member));
            }
            await this.voiceChannel.permissionOverwrites.set(overwrites);
            await this.textChannel.permissionOverwrites.set(overwrites);
        }
        await this.voiceChannel.edit({ name: this.name, permissionOverwrites: overwrites });
        await this.textChannel.edit({ name: this.name, permissionOverwrites: overwrites });
        await this.welcomeMessage?.edit({ embeds: this.getEmbeds() });
        await this.save();
        this.getVoiceTemplate().save();
    }

    async save() {
        await VoiceDB.update(this.toJSON(), { where: { creatorId: this.creator.id, voiceId: this.voiceChannelId } });
    }

    async load(voiceChannelId = null) {
        if (voiceChannelId) this.voiceChannelId = voiceChannelId;
        const voice = await VoiceDB.findOne({ where: { voiceId: this.voiceChannelId } });
        if (!voice) return this;
        this.creator = this.guild.members.cache.get(voice.creatorId);
        this.name = this.voiceChannel.name;
        this.voiceChannelId = voice.voiceId;
        this.textChannelId = voice.textId;
        this.voiceChannel = await this.guild.channels.cache.get(this.voiceChannelId);
        this.textChannel = await this.guild.channels.cache.get(this.textChannelId);
        this.whitelist = voice.whitelist;
        this.members = new Map(voice.members.map(m => [m, this.guild.members.cache.get(m)]));

        if (!this.voiceChannel || !this.textChannel) {
            await this.delete();
            return this;
        }
        const welcomeId = voice.welcomeId;
        if (welcomeId) {
            this.welcomeMessage = await this.textChannel.messages.fetch(welcomeId);
        }
        this.creationTimestamp = this.voiceChannel.createdTimestamp;

        this.loaded = true;
        return this;
    }

    getEmbeds() {
        const memberString = [...this.members.keys()].map(id => userMention(id)).join('\n');
        const embed = new MessageEmbed()
            .setColor('GREEN')
            .setTitle(`${this.name} Created Successfully`)
            .setDescription(`This channel will persist for ${Voice.MINUTES} minutes or until everyone is gone, whichever comes last.\nKeep in mind, Administrators can still see all of these channels.\n` +
                'To edit this channel please use the `/voice edit` command. All changes will be saved for the next time you create a channel.\n' +
                'To add members to your whitelist/blacklist, right click them in Discord then click Apps then Add to Voice.')
            .addField('Creator', userMention(this.creator.id), true)
            .addField(`${this.whitelist ? 'Allowed' : 'Blocked'} People`, memberString ? memberString : 'None', true);
        return [embed];
    }

    async checkDelete() {
        if (!this.loaded) return false;
        const checkDate = new Date(this.creationTimestamp + 1000 * 60 * Voice.MINUTES);
        if (checkDate < Date.now()) {
            if (this.voiceChannel.members.size === 0) {
                await this.delete();
                return true;
            }
        }
        return false;
    }

    async delete() {
        await VoiceDB.destroy({ where: { creatorId: this.creator.id } });
        this.server.privateVoiceChannels.delete(this.textChannelId);
        if (this.voiceChannel) await this.voiceChannel.delete();
        if (this.textChannel) await this.textChannel.delete();
    }

    async addMember(member) {
        if (this.members.has(member.id) || member.id === this.guild.client.user.id) return false;
        const overwrites = [...this.voiceChannel.permissionOverwrites.cache.values()];
        overwrites.push(this.getPermission(member));
        await this.voiceChannel.permissionOverwrites.set(overwrites);
        await this.textChannel.permissionOverwrites.set(overwrites);
        this.members.set(member.id, member);
        if (this.welcomeMessage) {
            const embed = this.welcomeMessage.embeds[0];
            const listField = embed.fields.find(f => f.name === 'Allowed People' || f.name === 'Blocked People');
            const memberString = [...this.members.keys()].map(id => userMention(id)).join('\n');
            listField.value = memberString;
            await this.welcomeMessage.edit({ embeds: [embed] });
        }
        await this.save();
        this.getVoiceTemplate().save();
        return true;
    }

    async removeMember(member) {
        if (!this.members.has(member.id) || member.id === this.guild.client.user.id) return false;
        await this.voiceChannel.permissionOverwrites.delete(member.id);
        await this.textChannel.permissionOverwrites.delete(member.id);
        this.members.delete(member.id);
        if (this.welcomeMessage) {
            const embed = this.welcomeMessage.embeds[0];
            const listField = embed.fields.find(f => f.name === 'Allowed People' || f.name === 'Blocked People');
            let memberString = [...this.members.keys()].map(id => userMention(id)).join('\n');
            if (!memberString) memberString = 'None';
            listField.value = memberString;
            await this.welcomeMessage.edit({ embeds: [embed] });
        }
        await this.save();
        this.getVoiceTemplate().save();
        return true;
    }

    static getAllowPermission(id, type = 'member') {
        return {
            id,
            allow: ['VIEW_CHANNEL'],
            type,
        };
    }

    static getDenyPermission(id, type = 'member') {
        return {
            id,
            deny: ['VIEW_CHANNEL'],
            type,
        };
    }

    getPermission(member) {
        if (this.whitelist) {
            const allowPermission = Voice.getAllowPermission(member.id);
            return allowPermission;
        } else {
            const denyPermission = Voice.getDenyPermission(member.id);
            return denyPermission;
        }
    }

    getVoiceTemplate() {
        return new VoiceLast(this.creator, this.name, this.whitelist, this.members);
    }
}

class VoiceLast {
    constructor(member, name = '', whitelist = false, members = new Map()) {
        this.member = member;
        this.name = name;
        this.whitelist = whitelist;
        this.members = members;
    }

    toJSON() {
        return {
            creatorId: this.member.id,
            name: this.name,
            whitelist: this.whitelist,
            members: [...this.members.keys()],
        };
    }

    async load(member = this.member) {
        const data = await VoiceTemplate.findOne({ where: { creatorId: member.id } });
        if (!data) return this;
        this.name = data.name;
        this.whitelist = data.whitelist;
        const memList = data.members.map(m => [m, member.guild.members.cache.get(m)]);
        this.members = new Map(memList.filter(m => m[1]));
        return this;
    }

    async save() {
        const data = await VoiceTemplate.findOne({ where: { creatorId: this.member.id } });
        if (!data) {
            await VoiceTemplate.create(this.toJSON());
        } else {
            await data.update(this.toJSON());
        }
    }
}

module.exports = {
    Voice,
    VoiceLast,
};